packages/firestore/src/local/shared_client_state.ts

```diff
@@ -0,0 +1,422 @@
+
+  constructor(private persistenceKey: string, private clientKey: string) {
+    if (!WebStorageSharedClientState.isAvailable()) {
+      throw new FirestoreError(
+        Code.UNIMPLEMENTED,
+        'LocalStorage is not available on this platform.'
+      );
+    }
+    this.storage = window.localStorage;
+    this.storageKey = this.toLocalStorageClientKey(this.clientKey);
+    this.activeClients[this.clientKey] = new LocalClientState();
+    this.clientStateKeyRe = new RegExp(
+      `^${CLIENT_STATE_KEY_PREFIX}_${persistenceKey}_([^_]*)$`
+    );
+  }
+
+  /** Returns 'true' if LocalStorage is available in the current environment. */
+  static isAvailable(): boolean {
+    return typeof window !== 'undefined' && window.localStorage != null;
+  }
+
+  start(knownClients: ClientKey[]): void {
+    assert(!this.started, 'WebStorageSharedClientState already started');
+    window.addEventListener('storage', this.storageListener);
+
+    for (const clientKey of knownClients) {
+      const storageKey = this.toLocalStorageClientKey(clientKey);
+      const clientState = RemoteClientState.fromLocalStorageEntry(
+        clientKey,
+        this.storage.getItem(storageKey)
+      );
+      if (clientState) {
+        this.activeClients[clientState.clientKey] = clientState;
+      }
+    }
+
+    this.started = true;
+    this.persistState();
+  }
+
+  getMinimumGlobalPendingMutation(): BatchId | null {
+    let minMutationBatch = null;
+    objUtils.forEach(this.activeClients, (key, value) => {
+      if (minMutationBatch === null) {
+        minMutationBatch = value.minMutationBatchId;
+      } else {
+        minMutationBatch = Math.min(value.minMutationBatchId, minMutationBatch);
+      }
+    });
+
+    return minMutationBatch;
+  }
+
+  getAllActiveQueryTargets(): SortedSet<TargetId> {
+    let activeTargets = new SortedSet<TargetId>(primitiveComparator);
+    objUtils.forEach(this.activeClients, (key, value) => {
+      activeTargets = activeTargets.unionWith(value.activeTargetIds);
+    });
+    return activeTargets;
+  }
+
+  addLocalPendingMutation(batchId: BatchId): void {
+    this.localClientState.addPendingMutation(batchId);
+    this.persistState();
+  }
+
+  removeLocalPendingMutation(batchId: BatchId): void {
+    this.localClientState.removePendingMutation(batchId);
+    this.persistState();
+  }
+
+  addLocalQueryTarget(targetId: TargetId): void {
+    this.localClientState.addQueryTarget(targetId);
+    this.persistState();
+  }
+
+  removeLocalQueryTarget(targetId: TargetId): void {
+    this.localClientState.removeQueryTarget(targetId);
+    this.persistState();
+  }
+
+  shutdown(): void {
+    assert(
+      this.started,
+      'WebStorageSharedClientState.shutdown() called when not started'
+    );
+    window.removeEventListener('storage', this.storageListener);
+    this.storage.removeItem(this.storageKey);
+    this.started = false;
+  }
+
+  private handleStorageEvent(event: StorageEvent): void {
```

Comment:
Just FYI- It occurred to me that once you start triggering events / callbacks to SyncEngine on changes we're going to need to make sure we are on the AsyncQueue, so you'll probably need to make WebStorageSharedClientState have an AsyncQueue reference, etc.

Since JS is single-threaded and we're not calling back into any other client code here, it doesn't matter for now.